### DISCUSSION TITLE
Is it possible to keep table design with custom queries?

### DISCUSSION THREAD

**Discussion by Mozix** [Budibase/budibase#11821] [Help] [answered]
Hello, 
is it possible to keep the table generated by budibase designed? like most of the table features are lost when a custom query is used.

and views aren't supported in MySQL Databases because I tried to return the data directly hopping the table design would remain.

If there is a workaround or whatsoever I would be happy for it!

Cheers!

**Comment by ConorWebb96** [comment]
Hey @Mozix,

I'm not quite sure what you mean, which features are lost whenever you use a custom query? To my knowledge the only one that requires a little bit of extra effort is pagination. Aside from that everything else should pretty much work out of the box with customer queries.

Thanks in advance.

**Reply by Mozix** [reply]
Hey!

The features im talking about are things like when I make a relationship between 2 tables, the related rows get displayed in the main table as little gray boxes and if a limit is reached it would say +2 more

I used the group_concat method in my custom query to try and get the data displayed as budibase hopefully does.

Here are 2 Screenshots
![image002.png](https://github.com/Budibase/budibase/assets/37887158/0ab4916b-cfaa-4207-9019-dcf406321962)

The first one being the query made with budibase itself and it's relationship


![image003.png](https://github.com/Budibase/budibase/assets/37887158/815b7447-b40f-45c6-84d4-a0f448b92a4a)

SECOND picture is my own custom query with the group_concat method in which you can see the data not fully being displayed and no option tells me to enable such feature.


Cheers!

**Reply by ConorWebb96** [reply, marked as answer]
Hey @Mozix,

I've had a little play around with this. Initially I had a more complicated answer but thankfully theres a really simple solution to this.

All you need to do is amend your custom query slightly. Instead of using `GROUP_CONCAT` can't you instead use `JSON_ARRAYAGG`. This will return the data in the format that Budibase needs to output in the format shown in your first screenshot. Below is what i think you'd need based on the screenshots you've provided.

`JSON_ARRAYAGG(e.proben) AS samplenames`

**Example** + **output** whenever testing this for you.

<img width="1301" alt="Screenshot 2023-09-20 at 10 56 11" src="https://github.com/Budibase/budibase/assets/126772285/5831c482-ce24-436b-83a3-0587721194f5">

<img width="1510" alt="Screenshot 2023-09-20 at 10 56 37" src="https://github.com/Budibase/budibase/assets/126772285/0d50aed5-aff2-4a77-9cf6-59b47612c8d9">

I hope this helps.

**Reply by Mozix** [reply]
Afternoon, I appreciate your help

I have tried to implement your idea/solution the issue still remains, right now I get an array of the items but they are not designed? 

Is it possible to see the preview you get? I am getting an object:

{
"I'd": 1,
"name": "test", 
"probename": "[\"Probetest1\", \"Probetest2\"]"
} 
The jsonarrayagg seems to be working, the array is being properly returned but the design still fails to be displayed. 

In the table I see the array normally and if there's no data I get [null] 


What about budibase Version? Does that play a role? I'm running the 2.9.25
And latest is 2.10.10

I have exactly implemented the statement and I'm getting the data correctly but yea only the array is being returned and not the items as shown with the design given by budibase.


Thanks in advance!

**Reply by Mozix** [reply]
I have went ahead and updated the app to the latest version and still the table doesn't update properly as mentioned above.

**Reply by ConorWebb96** [reply]
Hey @Mozix,

I don't think that part of the codebase has had any changes recently between those version. (Although it wouldn't hurt to stay up-to-date) I'm on the latest version of 2.10.10 if that helps.

Strangely enough looking at the output of the field i've changed to an array seems to output as a string here;
```
{
  "DepartmentID": 1,
  "DepartmentName": "HR",
  "EmployeeNames": "John,Emily"
}
```
And the preview looks very different to what I'm seeing within the design area.
<img width="707" alt="Screenshot 2023-09-20 at 12 28 27" src="https://github.com/Budibase/budibase/assets/126772285/3256a23c-c3cc-43b8-ab89-c6a54a3fe2da">

Whereas if i look at the network request it shows up as an array similar to what you see in your object.
<img width="484" alt="Screenshot 2023-09-20 at 12 33 18" src="https://github.com/Budibase/budibase/assets/126772285/33b29598-d38f-44fd-81bd-54452be99d92">

And the output shows of the table looks like;
<img width="1511" alt="Screenshot 2023-09-20 at 12 41 55" src="https://github.com/Budibase/budibase/assets/126772285/dcae10d8-0f9b-4a28-aa75-7263dd61e00a">

> I have exactly implemented the statement and I'm getting the data correctly but yea only the array is being returned and not the items as shown with the design given by budibase.

Could you share some screenshots on the amends you've made? As well as a screenshot of the tables output now?

Thanks again.

**Reply by Mozix** [reply]
Hey, sure here are the Screenshots of my SQL Statement and the preview as well as the table:

![image001.png](https://github.com/Budibase/budibase/assets/37887158/6eec4aab-6339-4a15-81f4-5608e12d0e32)

![1695213232980_image002.png](https://github.com/Budibase/budibase/assets/37887158/c40c33ae-3d66-4653-8c0d-7e3b98dbb6d8)

I hope this helps us reach a great solution! 


Thanks in advance!

**Reply by ConorWebb96** [reply]
Hey @Mozix,

If you don't mind would you be able to email your MySQL schema dump (It doesn't have to include data, I just want to create your tables with the exact same settings.) and app export to support@budibase.com, so I can take a closer look. 

I'm not seeing any difference here and hoping, replicating with your settings may help me find the issue you are having!

Thanks.

**Reply by Mozix** [reply]
Hey @ConorWebb96 

I have sent the E-Mail!

Cheers

**Reply by ConorWebb96** [reply]
Hey @Mozix,

I have incorporated your db schema and everything seems to be working fine for me? I can't replicate the issue you are having.

```
SELECT
    d.taskId,
    d.taskname,
    JSON_ARRAYAGG(e.samplename) AS samplename
FROM
    task d
LEFT JOIN
    sample e ON d.taskId = e.taskId
GROUP BY
    d.taskId, d.taskname;
```

**Output**
<img width="1512" alt="Screenshot 2023-09-21 at 08 00 37" src="https://github.com/Budibase/budibase/assets/126772285/af96b16f-81e7-42a5-8a07-a6a1ebb85ea1">

Settings of the entire query;
<img width="683" alt="Screenshot 2023-09-21 at 08 03 24" src="https://github.com/Budibase/budibase/assets/126772285/0f76cb8d-b0be-4674-9b60-9d3ead1a0443">
<img width="1096" alt="Screenshot 2023-09-21 at 08 03 16" src="https://github.com/Budibase/budibase/assets/126772285/c3bfa0a9-970b-4773-a323-cb104d8dc8f6">
<img width="734" alt="Screenshot 2023-09-21 at 08 03 42" src="https://github.com/Budibase/budibase/assets/126772285/08aabc82-a11c-468a-a281-1f141e6e7838">
<img width="677" alt="Screenshot 2023-09-21 at 08 03 33" src="https://github.com/Budibase/budibase/assets/126772285/5561e8c5-e9fb-475c-8c60-ff6ed298871b">

I hope this helps.

**Reply by Mozix** [reply]
Hey,

I am trying to figure out the core issue and I have a couple questions:

You are using MariaDB correct? If so which version are you running?

Do you have some features available on your budibase end that I don't have? Any settings or plugins whatsoever?

I have also noticed under results > schema the column has the value null? For me it's Text is that normal? Or is it supposed to be null to get the formatting?

Also how does the Json_arrayagg function look like if you run it normally on whatever sql executer you have?


Thanks for being patient!

**Reply by ConorWebb96** [reply]
Sure thing ill go through each of these and answer your questions;

- I am using MySQL as you mentioned this in your initial post, server version - 8.0.32
- This is a free tier self-host installation, so i should have all the features or less if you are on a paid plan.
- The value returning null wouldn't affect anything, you can do some frontend logic to avoid this showing up.
- I never ran it previous in an sql executer. I have now done it using a plugin within VScode and heres the output I'm receiving
<img width="534" alt="Screenshot 2023-09-21 at 10 20 09" src="https://github.com/Budibase/budibase/assets/126772285/cb5480f0-d566-4802-b5b2-6cc5b4a2a6d1">

Whats your samplename value type set to? Mine seems to be set to null, is yours text by any chance? If so there may need to be a potential enhancement here to account for arrays from custom queries.
<img width="656" alt="Screenshot 2023-09-21 at 10 25 13" src="https://github.com/Budibase/budibase/assets/126772285/b5fcf578-6838-4394-98b1-8d495719783f">

This could be specific to Maria DB.

**Reply by Mozix** [reply]
Hey,

So I have went ahead and setup a quick MySQL Server for test purposes.... And it worked? It's good and all but
When im gonna deploy the app I won't be able/allowed to use MySQL Server but only MariaDB.

IS the JSON_ARRAYAGG function different or what's wrong here?
Or does Budibase not accept any MariaDB calls?

Or will Budibase eventually support it?

Or maybe a workaround for the time being?

Nonetheless thank you so much for your time, patience and efforts I will probably set this as answered because technically it has been answered.
